### PR TITLE
feat: bring over buildQuoteTx logic from platform-shared

### PR DESCRIPTION
**Pull request by DaoDev44**
feat(ZrxSwapper): Add `buildQuoteTx` method to `ZrxSwapper`

- Refactor private methods to helper functions for easier testing
- add tests

closes #71

### CODE REVIEW

File: packages/swapper/src/swappers/zrx/buildQuoteTx/buildQuoteTx.ts [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+    if (e.status === 400) {
+      return {
+        sellAsset,
+        buyAsset,
+        success: false,
+        statusCode: e.body?.code || -1,
+        statusReason: e.body?.reason || 'Unknown Client Failure'
+      }
+    } else if (e.status === 500) {
+      // TODO: Handle error
```
Comment (toshiSat) [inline packages/swapper/src/swappers/zrx/buildQuoteTx/buildQuoteTx.ts:188]: This is probably leftover from platform-shared, I wonder if it should be combined with the statement above so it catches both of them.  Also, is e.status right here? It seems weird that it's grabbing the code from e.body?.code, but checking e.status

**Review by toshiSat — APPROVED** [review]
One comment on error handling, but other than that lgtm

File: packages/swapper/src/swappers/zrx/buildQuoteTx/buildQuoteTx.ts (hunk 1 above)
Comment (DaoDev44) [reply packages/swapper/src/swappers/zrx/buildQuoteTx/buildQuoteTx.ts:188]: Good catch. Yea I'm not exactly sure as to why these two cases are separated, but it does make sense to combine them unless there is some edge case where we need it. If we end up needed to separate the two cases, we can at that point.

I also think it's odd that `e.status` is being checked. My assumption is that we are getting a status back from the error message, but that could be wrong. Until we are able to fully test out the implementation, my thought was to keep it `as-is` until we can confidently make that decision. Thoughts?

File: packages/swapper/src/manager/SwapperManager.test.ts [hunk 2]
```diff
@@ -1,9 +1,16 @@
 /* eslint-disable @typescript-eslint/ban-ts-comment */
+import Web3 from 'web3'
 import { SwapperType } from '../api'
 import { ThorchainSwapper, ZrxSwapper } from '../swappers'
 import { SwapperManager } from './SwapperManager'
+import { ChainAdapterManager } from '@shapeshiftoss/chain-adapters'
 
 describe('SwapperManager', () => {
+  const zrxSwapperDeps = {
+    web3: ({} as unknown) as Web3,
```
Comment (kaladinlight) [inline packages/swapper/src/manager/SwapperManager.test.ts:10]: This should work as well and looks a little cleaner imo
```suggestion
    web3: <Web3>{},
```

**Review by 0xdef1cafe — CHANGES_REQUESTED** [review]
half way through just requesting these changes so they don't get lost

File: packages/swapper/src/swappers/zrx/buildQuoteTx/buildQuoteTx.ts [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+    )
+  }
+
+  // TODO: (ryankk) Remove the type cast when we unify ChainIdentifier and ChainTypes
+  const adapter: ChainAdapter = adapterManager.byChain(
+    (buyAsset.chain as unknown) as ChainIdentifier
+  )
+  const receiveAddress = await adapter.getAddress({ wallet, path: DEFAULT_ETH_PATH })
+
+  const slippagePercentage = slippage
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/zrx/buildQuoteTx/buildQuoteTx.ts:89]: do we wanna put some sanity bounds around slippage here with an optional param to override sanity checks so people don't get rekt with 50% slippage?

File: packages/swapper/src/swappers/zrx/buildQuoteTx/buildQuoteTx.ts [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+    quote.allowanceGrantRequired = allowanceRequired.gt(0)
+    if (quote.allowanceGrantRequired) {
+      quote.feeData = {
+        ...quote.feeData,
+        approvalFee: new BigNumber(APPROVAL_GAS_LIMIT).multipliedBy(data.gasPrice || 0).toString()
+      }
+    }
+    return quote
+  } catch (e) {
+    if (e.status === 400 || e.status === 500) {
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/zrx/buildQuoteTx/buildQuoteTx.ts:179]: refactor to use error handling from getQuote

File: packages/swapper/src/swappers/zrx/utils/applyAxiosRetry.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import { AxiosInstance, AxiosRequestConfig } from 'axios'
+import { attach, RetryConfig } from 'retry-axios'
+
+type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/zrx/utils/applyAxiosRetry.ts:4]: did we reinvent a wheel here or is this doing something different?

https://www.typescriptlang.org/docs/handbook/utility-types.html#omittype-keys

File: packages/swapper/src/swappers/zrx/utils/constants.ts [hunk 6]
```diff
@@ -11,3 +11,5 @@ export const MAX_GAS_PRICE = '1000000000000' // 1000 GWei
 export const MAX_ZRX_TRADE = '100000000000000000000000000'
 export const DEFAULT_SOURCE = [{ name: '0x', proportion: '1' }]
 export const DEFAULT_SLIPPAGE = '3.0' // 3%
+export const DEFAULT_ETH_PATH = `m/44'/60'/0'/0/0`
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/zrx/utils/constants.ts:14]: can we add a todo above this to remove once with add and make this optional in chain adapters?

File: packages/swapper/src/swappers/zrx/utils/helpers/helpers.test.ts [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+          }))
+        }
+      }))
+
+      expect(
+        await getAllowanceRequired({ quote: quoteInput, web3: web3Instance, erc20AllowanceAbi })
+      ).toEqual(new BigNumber(quoteInput.sellAmount))
+    })
+
+    it('should thow error if allowanceOnChain is undefined', async () => {
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/zrx/utils/helpers/helpers.test.ts:77]: typo

File: packages/swapper/src/swappers/zrx/utils/helpers/helpers.test.ts [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+        }
+      }))
+
+      expect(await getAllowanceRequired({ quote, web3: web3Instance, erc20AllowanceAbi })).toEqual(
+        new BigNumber(900)
+      )
+    })
+  })
+
+  describe('getDeps', () => {
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/zrx/utils/helpers/helpers.test.ts:135]: kill this and the kung fu utility and just do `this.deps = deps` in constructor of `buildQuote`

File: packages/swapper/src/swappers/zrx/utils/helpers/helpers.ts [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+  if (!allowanceOnChain) {
+    throw new SwapError(
+      `No allowance data for ${quote.allowanceContract} to ${quote.receiveAddress}`
+    )
+  }
+  const allowanceRequired = new BigNumber(quote.sellAmount || 0).minus(allowanceOnChain)
+  return allowanceRequired.lt(0) ? new BigNumber(0) : allowanceRequired
+}
+
+export function getDeps<
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/zrx/utils/helpers/helpers.ts:55]: kill this black magic

File: packages/swapper/src/swappers/zrx/utils/test-data/setupSwapQuote.ts [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    precision: 18,
+    tokenId: '0xc770eefad204b5180df6a14ee197d99d808ee52d',
+    contractType: ContractTypes.ERC20,
+    color: '#FFFFFF',
+    secondaryColor: '#FFFFFF',
+    icon: 'https://assets.coincap.io/assets/icons/fox@2x.png',
+    sendSupport: true,
+    receiveSupport: true,
+    symbol: 'FOX'
+  } as unknown as Asset
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/zrx/utils/test-data/setupSwapQuote.ts:18]: add some todos to remove these when we have the `lib/types` package with unified `ChainTypes` and `ChainIdentifier`

**Review by 0xdef1cafe — CHANGES_REQUESTED** [review]

**Review by 0xdef1cafe — APPROVED** [review]

**Pull request merged**